policy/release/repo_ids.rego

```diff
@@ -0,0 +1,170 @@
+	result := lib.result_helper(rego.metadata.chain(), [problem])
+}
+
+data_problems contains reason if {
+	not known_repo_ids
+	reason := "The 'known_rpm_repositories' key was not found in the data"
+}
+
+data_problems contains reason if {
+	not is_array(known_repo_ids)
+	reason := "The 'known_rpm_repositories' data is not an array"
+}
+
+data_problems contains reason if {
+	count(known_repo_ids) == 0
+	reason := "The 'known_rpm_repositories' data is empty"
+}
+
+data_problems contains reason if {
+	is_array(known_repo_ids)
+	non_strings := [r |
+		some r in known_repo_ids
+		not is_string(r)
+	]
+	count(non_strings) > 0
+	reason := "The 'known_rpm_repositories' data is not a list of strings"
+}
+
+# See https://github.com/release-engineering/rhtap-ec-policy/tree/main/data
+# This list of repo_ids is not under the 'rule_data' key, so we don't use lib.rule_data.
+known_repo_ids := data.known_rpm_repositories
+
+repo_id_problems contains reason if {
+	bad_items := all_rpm_purls - raw_purls(all_parsed_rpm_purls)
+	count(bad_items) > 0
+
+	reason := sprintf(
+		"Some rpm components in the SBOM have purls that could not be parsed: %s",
+		[truncated_list_to_string(bad_items)],
+	)
+}
+
+repo_id_problems contains reason if {
+	bad_items := all_parsed_rpm_purls - all_parsed_purls_with_repo_ids
+	count(bad_items) > 0
+
+	reason := sprintf(
+		"Some rpm components in the SBOM did not specify a repository_id value in their purl: %s",
+		[truncated_list_to_string(raw_purls(bad_items))],
+	)
+}
+
+repo_id_problems contains reason if {
+	bad_items := all_parsed_purls_with_repo_ids - all_parsed_purls_with_known_repo_ids
+	count(bad_items) > 0
+
+	reason := sprintf(
+		"Some rpm components in the SBOM specify an unknown or disallowed repository_id: %s",
+		[truncated_list_to_string(raw_purls(bad_items))],
+	)
+}
+
+all_parsed_purls_with_known_repo_ids contains p if {
+	some p in all_parsed_purls_with_repo_ids
+	purl_repo_id(p.parsed) in known_repo_ids
+}
+
+all_parsed_purls_with_repo_ids contains p if {
+	some p in all_parsed_rpm_purls
+	purl_repo_id(p.parsed) # any repo id
+}
+
+# Keep the raw and parsed purl in a little struct-like
+all_parsed_rpm_purls contains p if {
+	some purl in all_rpm_purls
+	ec.purl.is_valid(purl)
+	p := {
+		"raw": purl,
+		"parsed": ec.purl.parse(purl),
+	}
+}
+
+all_rpm_purls contains p if {
+	some sbom in all_sboms
+	some component in sbom.components
+	p := component.purl
+
+	# I'm assuming this is faster than parsing it and checking the type
+	startswith(p, "pkg:rpm")
+}
+
+# In future there will be SPDX sboms also
+all_sboms := lib.sbom.cyclonedx_sboms
+
+# Extract a repo id from a parsed purl
+purl_repo_id(parsed_purl) := purl_qualifier("repository_id", parsed_purl)
+
+# Extract a named qualifier
+purl_qualifier(key, parsed_purl) := result if {
+	some qualifier in parsed_purl.qualifiers
+	qualifier.key == key
+	result := qualifier.value
+}
+
+# Avoid including thousands of bad purls in the violation reason
+max_bad_purls := 10
```

Comment:
This seems oddly specific. Let all the failures show up so users can tackle them as needed.

Reply:
Are you worried about showing (say) 5000 violations?

Reply:
Because every purl is malformed? That seems like a bigger problem.

Reply:
My thinking is: if you do have this problem, it's better if we don't add a 9000 screen EC task log on top of problem pile. :grin: